// ionic/commands/info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const format_1 = require("@ionic/cli-framework/utils/format");
const string_1 = require("@ionic/cli-framework/utils/string");
const lodash = require("lodash");
const color_1 = require("../lib/color");
const command_1 = require("../lib/command");
const INFO_GROUPS = ['ionic', 'capacitor', 'cordova', 'utility', 'system', 'environment'];
class InfoCommand extends command_1.Command {
    async getMetadata() {
        return {
            name: 'info',
            type: 'global',
            summary: 'Print project, system, and environment information',
            description: `
This command is an easy way to share information about your setup. If applicable, be sure to run ${color_1.input('ionic info')} within your project directory to display even more information.
      `,
            options: [
                {
                    name: 'json',
                    summary: 'Print system/environment info in JSON format',
                    type: Boolean,
                },
            ],
        };
    }
    async run(inputs, options) {
        const { json } = options;
        if (json) {
            process.stdout.write(JSON.stringify(await this.env.getInfo()));
        }
        else {
            const results = await this.env.getInfo();
            const groupedInfo = new Map(INFO_GROUPS.map((group) => [group, results.filter(item => item.group === group)]));
            const sortInfo = (a, b) => {
                if (a.key[0] === '@' && b.key[0] !== '@') {
                    return 1;
                }
                if (a.key[0] !== '@' && b.key[0] === '@') {
                    return -1;
                }
                return string_1.strcmp(a.key.toLowerCase(), b.key.toLowerCase());
            };
            const projectPath = this.project && this.project.directory;
            const splitInfo = (ary) => ary
                .sort(sortInfo)
                .map((item) => [`   ${item.key}${item.flair ? ' ' + color_1.weak('(' + item.flair + ')') : ''}`, color_1.weak(item.value) + (item.path && projectPath && !item.path.startsWith(projectPath) ? ` ${color_1.weak('(' + item.path + ')')}` : '')]);
            const format = (details) => format_1.columnar(details, { vsep: ':' });
            if (!projectPath) {
                this.env.log.warn('You are not in an Ionic project directory. Project context may be missing.');
            }
            this.env.log.nl();
            for (const [group, info] of groupedInfo.entries()) {
                if (info.length > 0) {
                    this.env.log.rawmsg(`${color_1.strong(`${lodash.startCase(group)}:`)}\n\n`);
                    this.env.log.rawmsg(`${format(splitInfo(info))}\n\n`);
                }
            }
        }
    }
}
exports.InfoCommand = InfoCommand;
